### PR TITLE
Heat map representations for latent space

### PR DESCRIPTION
**Pull request by alecokas**
I have introduced functionality to produce a heat map for the mean of each latent dimension in the latent space. This functionality is of course heavily dependent on PR #9. 

I have tested that heat maps are produced with MNIST and dSprites (by cloning Bart's fork while we wait for the requested changes to that PR).

### CODE REVIEW

File: run_experiment.py [hunk 1]
```diff
@@ -1,24 +1,42 @@
...
 
-        for (test_data, _) in test_loader:
-            return test_data
+        for batch_idx, (test_data, label) in enumerate(data_loader):
+            if num_samples == batch_idx:
+                break
+            if batch_idx == 0:
+                data = test_data
+            else:
+                data = torch.cat(tensors=(data, test_data))
```
Comment (YannDubs) [inline run_experiment.py]: don't concatenate in loops : that's heavy.
Append to a list at every iteration (list append is O(1) in python)
then `torch.cat(list)`at the end of the loop.

If you want to be pythonic use `itertools.takewhile` and list comprehensions. But i's not curcuial at all

Comment (alecokas) [reply run_experiment.py]: I've just looked at the ```torch.cat()``` documentation in detail. I see the functionality is set up for multiple concatenations . Thanks for pointing this out!

File: run_experiment.py [hunk 2]
```diff
@@ -1,24 +1,42 @@
...
     with open('experiments/{}/specs.json'.format(experiment_name)) as spec_file:
         spec_data = json.load(spec_file)
         dataset_name = spec_data['dataset']
 
-        _, test_loader = get_dataloaders(batch_size=num_samples, dataset=dataset_name, shuffle=True)
+        data_loader = get_dataloaders(batch_size=batch_size, dataset=dataset_name, shuffle=shuffle)
 
-        for (test_data, _) in test_loader:
-            return test_data
+        for batch_idx, (test_data, label) in enumerate(data_loader):
```
[the remark this answers is not in the document]
Comment (alecokas) [reply run_experiment.py]: Noted

File: viz/visualize.py [hunk 3]
```diff
@@ -23,6 +24,45 @@ def __init__(self, model, save_images=True):
...
+
+            heat_map = torch.zeros([1, 1, heat_map_size[0], heat_map_size[1]], dtype=torch.int32)
+
+            for latent_dim in range(means.shape[1]):
+                for y_posn in range(32):
+                    for x_posn in range(32):
+                        heat_map[0, 0, x_posn, y_posn] = means[32 * y_posn + x_posn, latent_dim]
+
+                if self.save_images:
+                    heat_map_name = filename[:-4] + '-{}'.format(latent_dim) + filename[-4:]
```
[the remark this answers is not in the document]
Comment (alecokas) [reply viz/visualize.py]: Yeah, you're right

File: viz/visualize.py [hunk 4]
```diff
@@ -23,6 +24,45 @@ def __init__(self, model, save_images=True):
...
+        # Pass data through VAE to obtain reconstruction
+        with torch.no_grad():
+            input_data = data.to(self.device)
+            _, latent_dist = self.model(input_data)
+            means = latent_dist[1]
+
+            heat_map = torch.zeros([1, 1, heat_map_size[0], heat_map_size[1]], dtype=torch.int32)
+
+            for latent_dim in range(means.shape[1]):
+                for y_posn in range(32):
```
[the remark this answers is not in the document]
Comment (alecokas) [reply viz/visualize.py]: So true, an oversight on my part.

File: run_experiment.py [hunk 5]
```diff
@@ -1,24 +1,42 @@
...
+            if num_samples == batch_idx:
+                break
+            if batch_idx == 0:
+                data = test_data
+            else:
+                data = torch.cat(tensors=(data, test_data))
+        return data
 
+def sprite_location():
```
[the remark this answers is not in the document]
Comment (alecokas) [reply run_experiment.py]: I thought it may be useful down the line so I kept it in. I can remove it though.

**Pull request merged**